### ISSUE TITLE
Investigate NPE while retryWithH1C for integration test

### ISSUE THREAD

**Issue by jrhee17** [line/armeria#2603]
While running a unit test for proxy clients based on the current master branch, I found a test started to fail in my local env. (I expect CI should also be failing)

**Relevant Test** `ProxyClientIntegrationTest.testProxy_protocolUpgrade_notSharableExceptionNotThrown`

It seems like we can't guarantee `channel.remoteAddress` is non-null.
(I can investigate later.. but probably after work.     I would appreciate if anyone else can also take a look)

```
13:51:22.147 [armeria-common-worker-nio-9-1] WARN  c.l.a.client.HttpSessionHandler - [id: 0xbfe17139, L:0.0.0.0/0.0.0.0:49458] Unexpected suppressed exception:
com.linecorp.armeria.common.ClosedSessionException: null
Caused by: java.lang.NullPointerException: remoteAddress
	at io.netty.util.internal.ObjectUtil.checkNotNull(ObjectUtil.java:33)
	at io.netty.bootstrap.Bootstrap.connect(Bootstrap.java:137)
	at com.linecorp.armeria.client.HttpChannelPool.connect(HttpChannelPool.java:416)
	at com.linecorp.armeria.client.HttpSessionHandler.channelInactive(HttpSessionHandler.java:320)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:262)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:248)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:241)
	at io.netty.handler.codec.ByteToMessageDecoder.channelInputClosed(ByteToMessageDecoder.java:386)
	at io.netty.handler.codec.ByteToMessageDecoder.channelInactive(ByteToMessageDecoder.java:351)
	at io.netty.handler.codec.http2.Http2ConnectionHandler.channelInactive(Http2ConnectionHandler.java:415)
	at com.linecorp.armeria.client.Http2ClientConnectionHandler.channelInactive(Http2ClientConnectionHandler.java:85)
```

**Comment by minwoox** [comment]
Thanks, this will be fixed by https://github.com/line/armeria/pull/2597/files#diff-a3214924862f187ed8ecbc1116282611R114

**Comment by minwoox** [comment]
Closed by #2597
